packages/graph-explorer/src/core/StateProvider/localDb.ts

```diff
@@ -0,0 +1,223 @@
+  await saveEntriesToDb(entries, localDb);
+  await removeExistingEntries(localDb);
+  await removePrefixFromRestoredEntries(localDb);
+
+  const appliedBackupState = await createBackupData(localDb);
+  logger.debug("Backup restored to local db", appliedBackupState);
+}
+
+export interface LocalDb {
+  keys(): Promise<string[]>;
+  clear(): Promise<void>;
+  getItem<T>(key: string): Promise<T | null>;
+  setItem<T>(key: string, value: T): Promise<T>;
+  removeItem(key: string): Promise<void>;
+}
+
+const SerializedBackupSchema = z.object({
+  backupSource: z.string(),
+  backupSourceVersion: z.string(),
+  backupVersion: z.string(),
+  backupTimestamp: z.coerce.date(),
+  data: z.record(z.string(), z.unknown()),
+});
+
+export type SerializedBackup = z.infer<typeof SerializedBackupSchema>;
+
+export async function createBackupData(localDb: LocalDb) {
+  await removeRestoredEntries(localDb);
+  const keys = await localDb.keys();
+  let backup = {};
+
+  logger.debug(`Exporting config data from localForage`, keys);
+
+  for (const key of keys) {
+    const value = await localDb.getItem(key);
+    logger.debug(`Adding ${key} to backup`, value);
+
+    backup = {
+      ...backup,
+      [`${key}`]: value,
+    };
+  }
+
+  const result: SerializedBackup = {
+    backupSource: "Graph Explorer",
+    backupSourceVersion: __GRAPH_EXP_VERSION__,
+    backupVersion: "1.0",
+    backupTimestamp: new Date(),
+    data: backup,
+  };
+
+  logger.debug(`Backup created`, result);
+
+  return result;
+}
+
+export async function removeRestoredEntries(localDb: LocalDb) {
+  const keys = (await localDb.keys()).filter(hasRestoredPrefix);
+
+  logger.log(`Found ${keys.length} restored entries to be removed.`);
+
+  for (const key of keys) {
+    logger.debug("Removing entry", key);
+    await localDb.removeItem(key);
+  }
+}
+
+export async function removeExistingEntries(localDb: LocalDb) {
+  const keys = (await localDb.keys()).filter(missingRestoredPrefix);
+
+  logger.log(`Found ${keys.length} existing entries to be removed.`);
+
+  for (const key of keys) {
+    logger.debug("Removing entry", key);
+    await localDb.removeItem(key);
+  }
+}
+
+const restoredEntryPrefix = "RESTORED-ENTRY-";
+
+function applyRestorePrefixToEntries(data: Record<string, unknown>) {
+  return Object.entries(data).map(
+    ([key, value]) => [addRestoredPrefix(key), value] as const
+  );
+}
+
+export function addRestoredPrefix(key: string) {
+  return `${restoredEntryPrefix}${key}`;
+}
+
+export function removeRestoredPrefix(key: string) {
+  return key.replace(restoredEntryPrefix, "");
+}
+
+export function hasRestoredPrefix(key: string) {
+  return key.startsWith(restoredEntryPrefix);
+}
+
+export function missingRestoredPrefix(key: string) {
+  return !hasRestoredPrefix(key);
+}
+
+export async function saveEntriesToDb(
+  entries: (readonly [string, unknown])[],
+  localDb: LocalDb
+) {
+  for (const [key, value] of entries) {
```

Comment:
Shall we apply a more concise syntax (and also possibly optimize performance by parellizing?) - 

Example only:
```
await Promise.all(entries.map(([key, value]) => localDb.setItem(key, value)));
```

Reply:
I'd prefer to minimize optimizations until necessary and lean in to more readable code. But if we find the backup/restore process is slow for some users, this is a clear win.